src/ripple/ledger/impl/Directory.cpp

```diff
@@ -0,0 +1,226 @@
+        Keylet const& key)
+    : view_(&view)
+    , root_(key)
+    , sle_(view_->read(root_))
+
+{
+    if (sle_ == nullptr)
+        throw std::runtime_error("Invalid AccountID");
+    indexes_ = &sle_->getFieldV256(sfIndexes);
+}
+
+#if defined(_MSC_VER) && _MSC_VER <= 1800
+Dir::Dir(Dir&& other)
+    : view_(std::move(other.view_))
+    , root_(std::move(other.root_))
+    , sle_(std::move(other.sle_))
+{
+}
+
+Dir&
+Dir::operator=(Dir&& other)
+{
+    if (this != &other)
+    {
+        view_ = std::move(other.view_);
+        root_ = std::move(other.root_);
+        sle_ = std::move(other.sle_);
+    }
+    return *this;
+}
+
+#endif
+
+auto
+Dir::begin() const ->
+    iterator
+{
+    auto it = iterator(*view_, root_, root_);
+    it.sle_ = sle_;
+    if (! indexes_->empty())
+    {
+        it.indexes_ = indexes_;
+        it.it_ = std::begin(*indexes_);
+        it.index_ = *it.it_;
+    }
+
+    return it;
+}
+
+auto
+Dir::end() const  ->
+    iterator
+{
+    return iterator(*view_, root_, root_);
+}
+
+iterator
+Dir::find(uint256 const& page_key, uint256 const& sle_key) const
+{
+    auto it = iterator(*view_, root_, keylet::page(page_key, 0));
+    if (root_.key == page_key)
+    {
+        it.sle_ = sle_;
+        it.indexes_ = indexes_;
+    }
+    else
+    {
+        it.sle_ = view_->read(it.page_);
+        if (it.sle_ == nullptr)
+            return end();
+        it.indexes_ = &it.sle_->getFieldV256(sfIndexes);
+    }
+
+    it.it_ = std::find(std::begin(*it.indexes_),
+        std::end(*it.indexes_), sle_key);
+    if (it.it_ == std::end(*it.indexes_))
+        return end();
+
+    it.index_ = *it.it_;
+    return it;
+}
+
+#if defined(_MSC_VER) && _MSC_VER <= 1800
+iterator::iterator(iterator&& other)
+    : view_(std::move(other.view_))
+    , root_(std::move(other.root_))
+    , page_(std::move(other.page_))
+    , index_(std::move(other.index_))
+    , cache_(std::move(other.cache_))
+    , sle_(std::move(other.sle_))
+    , indexes_(std::move(other.indexes_))
+    , it_(std::move(other.it_))
+{
+}
+
+iterator&
+iterator::operator=(iterator&& other)
+{
+    if (this != &other)
+    {
+        view_ = std::move(other.view_);
+        root_ = std::move(other.root_);
+        page_ = std::move(other.page_);
+        index_ = std::move(other.index_);
+        cache_ = std::move(other.cache_);
+        sle_ = std::move(other.sle_);
+        indexes_ = std::move(other.indexes_);
+        it_ = std::move(other.it_);
+    }
+    return *this;
+}
+
+#endif
+
+iterator&
+iterator::operator=(iterator const& other)
+{
+    if (this != &other)
+    {
+        view_ = other.view_;
+        root_ = other.root_;
+        page_ = other.page_;
+        index_ = other.index_;
+        cache_ = other.cache_;
+        sle_ = other.sle_;
+        indexes_ = other.indexes_;
+        it_ = other.it_;
+    }
+    return *this;
+}
+
+bool
+iterator::operator==(iterator const& other) const
+{
+    if (view_ == nullptr || other.view_ == nullptr)
```

Comment:
Whoops, yes it did pass, just with different wording than in the proposal (that's why I couldn't find it).  It only applies to forward iterators, not input.  But if you value-initialize two forward iterators, you can compare them for equality.